### ISSUE TITLE
(question) after assert all type not narrowed (list with potential Nones)

### ISSUE THREAD

**Issue by FrancescElies** [microsoft/pyright#4641]
**Describe the bug**
In the example below after `assert all(...)` type narrowing seems to have no effect.

**To Reproduce**
```python
# file: list_with_nones.py
# pyright: strict


def to_str(x: int) -> None | str:
    return str(x) if x % 2 else None


mylist = [1, 3]
texts = [to_str(x) for x in mylist]
assert all(x is not None for x in texts)
print("\n".join(texts))
```
```nu
❯ with-env { PYRIGHT_PYTHON_FORCE_VERSION: 'latest' } { pyright list_with_nones.py }
No configuration file found.
No pyproject.toml file found.
stubPath C:\s\mytest\typings is not a valid directory.
Assuming Python platform Windows
Searching for source files
Found 1 source file
pyright 1.1.294
C:\s\mytest\list_with_nones.py
  C:\s\mytest\list_with_nones.py:11:17 - error: Argument of type "list[str | None]" cannot be assigned to parameter "__iterable" of type "Iterable[str]" in function "join"
    "list[str | None]" is incompatible with "Iterable[str]"
      TypeVar "_T_co@Iterable" is covariant
        Type "str | None" cannot be assigned to type "str"
          Type "None" cannot be assigned to type "str" (reportGeneralTypeIssues)
  C:\s\mytest\list_with_nones.py:11:7 - error: Argument type is unknown
    Argument corresponds to parameter "values" in function "print" (reportUnknownArgumentType)
2 errors, 0 warnings, 0 informations
Completed in 0.652sec
```

**Expected behavior**
`texts` is narrowed from `list[str|None]` to `list[str]` but maybe I am missing something?

**Comment by erictraut** [comment]
Yes, this is by design. You're using a conditional expression that is not a supported type guard form. For a complete list of supported type guards, see [this documentation](https://github.com/microsoft/pyright/blob/main/docs/type-concepts.md#type-guards).

It's unlikely we'd ever support this form. Type guard forms are very specific, so this would require hard-coded knowledge of the `all` function and an argument that is a generator that takes the form `x is not None for x in y`. 

My recommendation is that you use a user-defined type guard (see [PEP 647](https://peps.python.org/pep-0647/)). Here's what that would look like:

```python
from typing import Iterable, TypeGuard, TypeVar

def to_str(x: int) -> None | str:
    return str(x) if x % 2 else None

mylist = [1, 3]
texts = [to_str(x) for x in mylist]

T = TypeVar("T")

def contains_no_none(val: Iterable[T | None]) -> TypeGuard[Iterable[T]]:
    return all(x is not None for x in val)

texts = filter(lambda x: x is not None, texts)
assert contains_no_none(texts)
print("\n".join(texts))
```
